### PR TITLE
build: Replace genisoimage with xorriso

### PR DESCRIPTION
**Pull request by dongcarl**
This is a redo of fanquake's https://github.com/bitcoin/bitcoin/pull/18151, which, aside from switching us from the deprecated `genisoimage` to the maintained `xorriso`, is also necessary for Guix to achieve determinism without using faketime.

> xorriso and its mkisofs/genisoimage emulation alter-ego xorrisofs are
> more maintained, and has the right toggles for us to achieve output
> determinism without using blunt tools like faketime.
> 
> In this commit, we use xorrisofs from the build environment rather than
> building it ourselves using depends. This is not necessary and can be
> changed in the future.
> 
> From wiki.debian.org/genisoimage?action=recall&rev=11 :
> 
> > The classical command line interface for production of ISO 9660
> > filesystem images is the option set established by program mkisofs.
> > For reasons of licensing and other problems with its author, Debian
> > ships a fork of mkisofs, called genisoimage, which was split off in
> > 2006 and then developed independently.
> >
> > Meanwhile, genisoimage gets no new features and not even bug fixes. It
> > is first choice only if its options -udf or -hfs are needed.
> >
> > Replacement in most uses cases, especially for bootable ISO 9660
> > filesystems, archiving, and backup, is xorrisofs which starts the -as
> > mkisofs emulation mode of program xorriso.

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK

**Comment by practicalswift** [comment]
Concept ACK: maintained is better than unmaintained :)

**Comment by hebasto** [comment]
Concept ACK.

**Comment by jonasschnelli** [comment]
Tested ACK f2cfb5c794c83a6e8771c8f08e5fa8aefcf269fc - Did a gitian build and verified the OSX disk image https://bitcoin.jonasschnelli.ch/gitian/build/347

**Comment by dongcarl** [comment]
Pushed b9d60ddefcf14c0ccf38c1a841c7558396e5e738 -> 75bfa827b30985ec7ca3b161a6dc0e1fb2f0432b, specify that we need to use the faketime-wrapped xorrisofs to `./configure` in Gitian

**Comment by fanquake** [comment]
Looks good. I tested with 48e1b68a2e757fee2dbcd724deee075385e797c5 on top of master (86f20071931b803b5f26ed8f685d98d4919fb7a7).

One thing I noticed is that the naming of the DMGs is now inconsistent. i.e a `make deploy` on macOS still produces `Bitcoin-Core.dmg`, whereas when cross-compiling you'll end up with `bitcoin-21.99.0.dmg` (along with the `.iso`). We should probably make this consistent. #20422 recently changed the name of the DMG produced on macOS (to `Bitcoin-Core.dmg over `Bitcoin-Qt.dmg`). Probably worth rebasing this branch so that you're working on top of those changes.

Will be nice if `xorrisofs` can do DMG compression in the future.

Note: we should also add a `*.iso` entry to .gitignore as part of this change.

**Comment by dongcarl** [comment]
Addressed fanquake's concerns. Here are the hashes for fdb76e62ce8d4d7a072d2e05895360a1cb7e7b6d

```
ee16ae845d4ea95842abd59e81b91c3a84924149c94c526c2f68d46f08b40cf3  bitcoin-fdb76e62ce8d-osx-unsigned.dmg
a96e1af388a4c1616489d4d52076d567c1ab33557bd0e56ea21fbd766739d33c  bitcoin-fdb76e62ce8d-osx-unsigned.tar.gz
c2691b955bf1d781f19c41908f9152264502a299b827b30d40049b8f567d35e6  bitcoin-fdb76e62ce8d-osx-unsigned.temp.iso
fc298e3f2f592398a386d9cf6eabaed8c41e7bf949636014ef78120430ef1351  bitcoin-fdb76e62ce8d-osx64.tar.gz
2e3e88d07f07bd42457ec22bf43fae6cfe51582faae50c0e0b0d41aae2c6d809  src/bitcoin-fdb76e62ce8d.tar.gz
```

File: Makefile.am [hunk 1]
```diff
@@ -135,8 +136,11 @@ $(APP_DIST_DIR)/Applications:
 
 $(APP_DIST_EXTRAS): $(APP_DIST_DIR)/$(OSX_APP)/Contents/MacOS/Bitcoin-Qt
 
-$(OSX_DMG): $(APP_DIST_EXTRAS)
-	$(GENISOIMAGE) -no-cache-inodes -D -l -probe -V "$(OSX_VOLNAME)" -no-pad -r -dir-mode 0755 -apple -o $@ dist
+$(OSX_TEMP_ISO): $(APP_DIST_EXTRAS)
```
Comment (theuni) [inline Makefile.am:139]: This could be .INTERMEDIATE, don't know why it wasn't before.

**Review by theuni — COMMENTED** [review]
Concept ACK. I never had much luck figuring out who the true upstreams were for these tools. Much better to use something supported.

**Comment by dongcarl** [comment]
Pushed fdb76e62ce8d4d7a072d2e05895360a1cb7e7b6d -> 25d39b61095d7c1ccbc5ec08a9c738aa56ade421: addressed https://github.com/bitcoin/bitcoin/pull/20470#discussion_r541049726

**Review by hebasto — APPROVED** [review]
ACK 25d39b61095d7c1ccbc5ec08a9c738aa56ade421, tested cross-compiling on Linux Mint 20 (x86_64), and made the gitian build:
```
616680767fe519b013f31a8c915719d15aeb45647929a2185e9d7d4b34ed057a  bitcoin-25d39b61095d-osx-unsigned.dmg
19e44ee5454510a7ce77b61ffdfd988f55ec1da8497b0c63803605601ef45ef0  bitcoin-25d39b61095d-osx-unsigned.tar.gz
ab639cd94bcb1ca168312927e1708de84a844be94e9672951d71de1446297f39  bitcoin-25d39b61095d-osx64.tar.gz
574fac8995ab9cb900aeb263e5b920d7ae3d1385523a4dd9cbe58c56184e1761  src/bitcoin-25d39b61095d.tar.gz
```

File: Makefile.am [hunk 2]
```diff
@@ -135,8 +136,12 @@ $(APP_DIST_DIR)/Applications:
 
 $(APP_DIST_EXTRAS): $(APP_DIST_DIR)/$(OSX_APP)/Contents/MacOS/Bitcoin-Qt
 
-$(OSX_DMG): $(APP_DIST_EXTRAS)
-	$(GENISOIMAGE) -no-cache-inodes -D -l -probe -V "$(OSX_VOLNAME)" -no-pad -r -dir-mode 0755 -apple -o $@ dist
+.INTERMEDIATE: $(OSX_TEMP_ISO)
+$(OSX_TEMP_ISO): $(APP_DIST_EXTRAS)
+	$(XORRISOFS) -D -l -V "$(OSX_VOLNAME)" -no-pad -r -dir-mode 0755 -o $@ dist
```
[the remark this answers is not in the document]
Comment (fanquake) [reply Makefile.am:141]: I'd rather we be explicit.

**Comment by dongcarl** [comment]
Pushed 25d39b61095d7c1ccbc5ec08a9c738aa56ade421 -> 7587d11ec959f15f469bd396d4ad2697729b4ccd: addressed @hebasto's comments, thanks!

```
6853d97647d9a77e9ddb7e7fdb117917fc64eff35b3b745c99c7f8a9af21dc91  bitcoin-7587d11ec959-osx-unsigned.dmg
e1546285a738ae3a7efea5f743e0e3493c00d3132fd65ff34938783c5ec7ca15  bitcoin-7587d11ec959-osx-unsigned.tar.gz
ad7074ef60772c80330a4fcaeada2ba56e2ce0e66ad29070d8463b3228da5b9e  bitcoin-7587d11ec959-osx64.tar.gz
5fea71337d11f362b086d4424a9254df7d2ec841c7fbac123ac68b71f49ea6f9  src/bitcoin-7587d11ec959.tar.gz
```

Maintainers: I think this is good to merge.

**Comment by sipa** [comment]
Concept ACK

**Comment by laanwj** [comment]
utACK 7587d11ec959f15f469bd396d4ad2697729b4ccd

File: Makefile.am [hunk 3]
```diff
@@ -135,8 +136,12 @@ $(APP_DIST_DIR)/Applications:
 
 $(APP_DIST_EXTRAS): $(APP_DIST_DIR)/$(OSX_APP)/Contents/MacOS/Bitcoin-Qt
 
-$(OSX_DMG): $(APP_DIST_EXTRAS)
-	$(GENISOIMAGE) -no-cache-inodes -D -l -probe -V "$(OSX_VOLNAME)" -no-pad -r -dir-mode 0755 -apple -o $@ dist
+.INTERMEDIATE: $(OSX_TEMP_ISO)
```
Comment (hebasto) [inline Makefile.am:139]: This apparently fires a warning:
```
$ ./autogen.sh 
...
Makefile.am:335: warning: .INTERMEDIATE was already defined in condition !BUILD_DARWIN, which is included in condition TRUE ...
Makefile.am:139: ... '.INTERMEDIATE' previously defined here
...
```

**Pull request merged**